modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -0,0 +1,777 @@
+     */
+    private void ensureCacheAbsent(String name) {
+        int id = CU.cacheId(name);
+
+        if (ctx.cache().cacheGroupDescriptors().containsKey(id) || ctx.cache().cacheDescriptor(id) != null) {
+            throw new IgniteIllegalStateException("Cache \"" + name +
+                "\" should be destroyed manually before perform restore operation.");
+        }
+    }
+
+    /**
+     * @param req Request to prepare cache group restore from the snapshot.
+     * @return Result future.
+     */
+    private IgniteInternalFuture<ArrayList<StoredCacheData>> prepare(SnapshotRestorePrepareRequest req) {
+        if (ctx.clientNode())
+            return new GridFinishedFuture<>();
+
+        try {
+            DiscoveryDataClusterState state = ctx.state().clusterState();
+
+            if (state.state() != ClusterState.ACTIVE || state.transition())
+                throw new IgniteCheckedException(OP_REJECT_MSG + "The cluster should be active.");
+
+            for (UUID nodeId : req.nodes()) {
+                ClusterNode node = ctx.discovery().node(nodeId);
+
+                if (node == null || !CU.baselineNode(node, state) || !ctx.discovery().alive(node)) {
+                    throw new IgniteCheckedException(
+                        OP_REJECT_MSG + "Required node has left the cluster [nodeId-" + nodeId + ']');
+                }
+            }
+
+            for (String grpName : req.groups())
+                ensureCacheAbsent(grpName);
+
+            opCtx = prepareContext(req);
+
+            SnapshotRestoreContext opCtx0 = opCtx;
+
+            if (opCtx0.dirs.isEmpty())
+                return new GridFinishedFuture<>();
+
+            // Ensure that shared cache groups has no conflicts.
+            for (StoredCacheData cfg : opCtx0.cfgs.values()) {
+                if (!F.isEmpty(cfg.config().getGroupName()))
+                    ensureCacheAbsent(cfg.config().getName());
+            }
+
+            if (log.isInfoEnabled()) {
+                log.info("Starting local snapshot restore operation [reqId=" + req.requestId() +
+                    ", snapshot=" + req.snapshotName() + ", group(s)=" + req.groups() + ']');
+            }
+
+            boolean updateMeta = ctx.localNodeId().equals(req.updateMetaNodeId());
+            Consumer<Exception> errHnd = (ex) -> opCtx.err.compareAndSet(null, ex);
+            BooleanSupplier stopChecker = () -> {
+                if (opCtx.err.get() != null)
+                    return true;
+
+                if (Thread.currentThread().isInterrupted()) {
+                    errHnd.accept(new IgniteInterruptedCheckedException("Thread has been interrupted."));
+
+                    return true;
+                }
+
+                return false;
+            };
+
+            GridFutureAdapter<ArrayList<StoredCacheData>> retFut = new GridFutureAdapter<>();
+
+            restoreAsync(opCtx0.snpName, opCtx0.dirs, updateMeta, stopChecker, errHnd).thenAccept(res -> {
```

Comment:
I think it would be more readable if you extract the functionality related to the `updateMeta` from this method.